catalyst/moodle-search_elastic
# Don't eat GuzzleExceptions when requests fail

**NB**: As we are testing situations in which Elasticsearch is unreachable, it is **not** necessary to have an Elasticsearch service running locally to complete testing.

# Prerequisites
1. Install [local_aws](https://github.com/catalyst/moodle-local_aws) (this plugin depends on it)
2. Set up global search in Moodle:
    1. Browse to "Site administration" > "Advanced features"
    2. Turn on "Global search"
    3. Press "Save changes"
3. Enable the Elastic search plugin
    1. Browse to "Site administration" > "Plugins" > "Search" > "Manage global search"
    2. Under "Select search engine" select  "Elastic"
    3. Press "Save changes"
4. Configure the Elastic search plugin
    1. Browse to "Site administration" > "Plugins" > "Search" > "Elastic" > "Plugin settings"
    2. Set "Hostname" to "localhost"
    3. **Verify** an error like "The configured elastic server at localhost:9200 returned the status code 503" is displayed (this is fine as we will hack the `is_server_ready` function to make the indexing scheduled task run)

# Testing
1. Hack `is_server_ready` in classes/engine.php so that it immediately returns true:
```
public function is_server_ready($stack = false) {
    return true;
    ...
````
2. Run the scheduled task:  `php admin/cli/scheduled_task.php --execute="\\core\\task\\search_index_task"`
3. **Verify** an error is present in the output like
```
Scheduled task failed: Global search indexing (core\task\search_index_task),cURL error 7: Failed to connect to localhost port 9200 after 0 ms: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)
```
4. Undo the hack from step 1.
5. Re-run the task
6. **Verify** the output looks like:
```
Scheduled task failed: Global search indexing (core\task\search_index_task),The search engine is not available. Please contact your administrator.
```

# Bonus
Set up an Elasticsearch server and verify that the plugin works correctly when properly configured

!fixes #117
